File: 2024/day04.py
import os
import sys

test = """\
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX"""

day = os.path.splitext(os.path.basename(__file__))[0]

TEST = 'test' in sys.argv
DEBUG = 'debug' in sys.argv

if TEST:
    data = test.splitlines()
else:
    data = open(day+'.txt').read().splitlines()

WIDTH = len(data[0])
HEIGHT = len(data)

# Find all occurrances of c.

def findall(c):
    return [(x,y) 
        for y,row in enumerate(data)
        for x,col in enumerate(row)
        if col==c]
               
dirs = ( (-1,-1), (-1,0), (-1,1),
         ( 0,-1),         (0,1),
         ( 1,-1), (1,0), (1,1))

def part1(data):
    winner = 0
    for x0,y0 in findall('X'):
        for dx,dy in dirs:
            x,y = x0,y0
            for c in 'MAS':
                x += dx
                y += dy
                if not (x in range(WIDTH) and y in range(HEIGHT) and data[y][x] == c):
                    break
            else:
                winner += 1
    return winner

def part2(data):
    winner = 0
    for x0,y0 in findall('A'):
        winner += (0 < x0 < WIDTH-1) and (0 < y0 < HEIGHT-1) and (
              (data[y0-1][x0-1] == 'M' and  data[y0+1][x0+1] == 'S') or 
              (data[y0-1][x0-1] == 'S' and  data[y0+1][x0+1] == 'M')
        ) and (
              (data[y0-1][x0+1] == 'M' and  data[y0+1][x0-1] == 'S') or 
              (data[y0-1][x0+1] == 'S' and  data[y0+1][x0-1] == 'M')
        )
    return winner

print("Part 1:", part1(data))
print("Part 2:", part2(data))

File: 2024/day04.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false

var test = `
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX`[1:]

//go:embed day04.txt
var live string

var WIDTH int = -1
var HEIGHT int = -1

// Find all occurrances of c.

type Point = tools.Point

func findall(data []string, c rune) []Point {
	var result []Point
	for y, row := range data {
		for x, col := range row {
			if col == c {
				result = append(result, Point{x, y})
			}
		}
	}
	return result
}

var directions []Point

func part1(data []string) int {
	winner := 0
	for _, point := range findall(data, 'X') {
		for _, dir := range directions {
			var pt = point
			winner++
			for _, c := range []byte("MAS") {
				pt.X += dir.X
				pt.Y += dir.Y
				if pt.InRange(WIDTH, HEIGHT) && data[pt.Y][pt.X] == c {
					winner--
					break
				}
			}
		}
	}
	return winner
}

func part2(data []string) int {
	winner := 0
	for _, pt := range findall(data, 'A') {
		if tools.Between(1, pt.X, WIDTH-1) &&
			tools.Between(1, pt.Y, HEIGHT-1) &&
			((data[pt.Y-1][pt.X-1] == 'M' && data[pt.Y+1][pt.X+1] == 'S') ||
				(data[pt.Y-1][pt.X-1] == 'S' && data[pt.Y+1][pt.X+1] == 'M')) &&
			((data[pt.Y-1][pt.X+1] == 'M' && data[pt.Y+1][pt.X-1] == 'S') ||
				(data[pt.Y-1][pt.X+1] == 'S' && data[pt.Y+1][pt.X-1] == 'M')) {
			winner++
		}
	}
	return winner
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)

	var data = strings.Split(input, "\n")
	WIDTH = len(data[0])
	HEIGHT = len(data)
	directions = []Point{
		{-1, -1}, {-1, 0}, {-1, 1},
		{0, -1}, {0, 1},
		{1, -1}, {1, 0}, {1, 1}}

	fmt.Println("Part 1:", part1(data))
	fmt.Println("Part 2:", part2(data))
}
